lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/security/FunctionAccessControlRule.java

```diff
@@ -0,0 +1,159 @@
+        this.schemaRegex = requireNonNull(schemaRegex, "schemaRegex is null");
+        this.functionRegex = requireNonNull(functionRegex, "functionRegex is null");
+        this.functionKinds = requireNonNull(functionKinds, "functionKinds is null");
+        checkState(!functionKinds.isEmpty(), "functionKinds cannot be empty, provide at least one function kind " + Arrays.toString(FunctionKind.values()));
+        globalScopeFunctionKind = functionKinds.contains(SCALAR) || functionKinds.contains(AGGREGATE) || functionKinds.contains(WINDOW);
+        //TODO: when every functions will be tied to connectors then remove this check
+        checkState(functionKinds.equals(Set.of(TABLE)) || schemaRegex.isEmpty(), "Cannot define schema for others function kinds than TABLE");
+    }
+
+    public boolean matches(String user, Set<String> roles, Set<String> groups, String functionName)
+    {
+        return globalScopeFunctionKind &&
+                userRegex.map(regex -> regex.matcher(user).matches()).orElse(true) &&
+                roleRegex.map(regex -> roles.stream().anyMatch(role -> regex.matcher(role).matches())).orElse(true) &&
+                groupRegex.map(regex -> groups.stream().anyMatch(group -> regex.matcher(group).matches())).orElse(true) &&
+                functionRegex.map(regex -> regex.matcher(functionName).matches()).orElse(true);
+    }
+
+    public boolean matches(String user, Set<String> roles, Set<String> groups, FunctionKind functionKind, SchemaRoutineName functionName)
+    {
+        return this.functionKinds.contains(functionKind) &&
+                userRegex.map(regex -> regex.matcher(user).matches()).orElse(true) &&
+                roleRegex.map(regex -> roles.stream().anyMatch(role -> regex.matcher(role).matches())).orElse(true) &&
+                groupRegex.map(regex -> groups.stream().anyMatch(group -> regex.matcher(group).matches())).orElse(true) &&
+                schemaRegex.map(regex -> regex.matcher(functionName.getSchemaName()).matches()).orElse(true) &&
+                functionRegex.map(regex -> regex.matcher(functionName.getRoutineName()).matches()).orElse(true);
+    }
+
+    public boolean canExecuteFunction()
+    {
+        return privileges.contains(EXECUTE) || canGrantExecuteFunction();
+    }
+
+    public boolean canGrantExecuteFunction()
+    {
+        return privileges.contains(GRANT_EXECUTE);
+    }
+
+    Optional<AnySchemaPermissionsRule> toAnySchemaPermissionsRule()
+    {
+        if (privileges.isEmpty() ||
+                //TODO: when every functions will be tied to connectors then remove this check
+                !containsTableFunctionKind()) {
+            return Optional.empty();
+        }
+        return Optional.of(new AnySchemaPermissionsRule(userRegex, roleRegex, groupRegex, schemaRegex));
+    }
+
+    boolean containsTableFunctionKind() //TODO: when every functions will be tied to connectors then remove this check
```

Comment:
there are 3 occurrences of this method. I think I'd prefer to inline it. 
Up to you.

Reply:
Agree, inline